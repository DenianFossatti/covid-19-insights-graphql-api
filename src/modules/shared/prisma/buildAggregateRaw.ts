import { Prisma } from '@prisma/client';
import { DefaultArgs } from '@prisma/client/runtime';

import { CustomGraphQLArgs } from '../../../types';
import { buildPrismaRangeWhere } from '../../shared/filters/DateRangeFilter';

export type BuildAggregateRawArgs = {
  args: CustomGraphQLArgs;
  groupBy?: Prisma.Covid_2022ScalarFieldEnum;
  aggregationType: 'avg' | 'countByMonth';
  match?: Record<string, any>;
};

export const buildAggregateRaw = ({
  args,
  aggregationType,
  groupBy,
  match,
}: BuildAggregateRawArgs): Prisma.covid_2022AggregateRawArgs<DefaultArgs> => {
  if ((aggregationType === 'avg' || aggregationType === 'countByMonth') && !groupBy)
    throw new Error('groupBy is required for aggregationType avg');

  const _match = {
    ...buildPrismaRangeWhere(args, true)?.where,
    ...match,
  };
  const _addFields =
    aggregationType === 'avg'
      ? undefined
      : {
          data_inclusao: { $toDate: '$data_inclusao' },
        };

  const _group =
    aggregationType === 'avg'
      ? { _id: null, field: { $avg: { $toInt: `$${groupBy}` } } }
      : {
          _id: {
            month: { $month: '$data_inclusao' },
            year: { $year: '$data_inclusao' },
          },
          field: { $avg: { $toInt: `$${groupBy}` } },
        };

  const _project =
    aggregationType === 'avg'
      ? { _id: 0, [groupBy!]: { $round: ['$field'] } }
      : {
          month: '$_id.month',
          year: '$_id.year',
          [groupBy!]: { $round: ['$field'] },
          _id: 0,
        };

  const _sort = aggregationType === 'avg' ? undefined : { year: 1, month: 1 };

  return {
    pipeline: [
      ...(_match ? [{ $match: _match }] : []),
      ...(_addFields ? [{ $addFields: _addFields }] : []),
      {
        $group: _group,
      },
      {
        $project: _project,
      },
      ...(_sort ? [{ $sort: _sort }] : []),
    ],
  };
};
