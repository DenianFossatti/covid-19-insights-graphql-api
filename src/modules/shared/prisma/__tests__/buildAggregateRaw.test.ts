import { buildAggregateRaw } from '../buildAggregateRaw';

describe('buildAggregateRaw', () => {
  it('should build aggregateRaw for avg', () => {
    const result = buildAggregateRaw({
      aggregationType: 'avg',
      groupBy: 'idade',
      args: {
        filters: {
          endDate: '2022-10-27T03:00:00.000Z',
          startDate: '2022-10-27T03:00:00.000Z',
        },
      },
    });

    expect(result).toMatchSnapshot();
  });

  it('should build aggregateRaw for countByMonth', () => {
    const result = buildAggregateRaw({
      aggregationType: 'countByMonth',
      groupBy: 'idade',
      args: {
        filters: {
          endDate: '2022-10-27T03:00:00.000Z',
          startDate: '2022-10-27T03:00:00.000Z',
        },
      },
    });

    expect(result).toMatchSnapshot();
  });

  it('should build aggregateRaw for countByMonth and match options', () => {
    const result = buildAggregateRaw({
      aggregationType: 'countByMonth',
      groupBy: 'idade',
      args: {
        filters: {
          endDate: '2022-10-27T03:00:00.000Z',
          startDate: '2022-10-27T03:00:00.000Z',
        },
      },
      match: {
        data_inclusao_obito: { $ne: null },
      },
    });

    expect(result).toMatchSnapshot();
  });
});
