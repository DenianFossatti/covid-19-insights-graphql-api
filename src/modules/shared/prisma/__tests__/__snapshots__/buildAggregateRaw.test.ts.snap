// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildAggregateRaw should build aggregateRaw for avg 1`] = `
{
  "pipeline": [
    {
      "$match": {
        "data_inclusao": {
          "$gte": "2022-10-27T03:00:00.000Z",
          "$lte": "2022-10-27T03:00:00.000Z",
        },
      },
    },
    {
      "$group": {
        "_id": null,
        "field": {
          "$avg": {
            "$toInt": "$idade",
          },
        },
      },
    },
    {
      "$project": {
        "_id": 0,
        "idade": {
          "$round": [
            "$field",
          ],
        },
      },
    },
  ],
}
`;

exports[`buildAggregateRaw should build aggregateRaw for countByMonth 1`] = `
{
  "pipeline": [
    {
      "$match": {
        "data_inclusao": {
          "$gte": "2022-10-27T03:00:00.000Z",
          "$lte": "2022-10-27T03:00:00.000Z",
        },
      },
    },
    {
      "$addFields": {
        "data_inclusao": {
          "$toDate": "$data_inclusao",
        },
      },
    },
    {
      "$group": {
        "_id": {
          "month": {
            "$month": "$data_inclusao",
          },
          "year": {
            "$year": "$data_inclusao",
          },
        },
        "field": {
          "$avg": {
            "$toInt": "$idade",
          },
        },
      },
    },
    {
      "$project": {
        "_id": 0,
        "idade": {
          "$round": [
            "$field",
          ],
        },
        "month": "$_id.month",
        "year": "$_id.year",
      },
    },
    {
      "$sort": {
        "month": 1,
        "year": 1,
      },
    },
  ],
}
`;

exports[`buildAggregateRaw should build aggregateRaw for countByMonth and match options 1`] = `
{
  "pipeline": [
    {
      "$match": {
        "data_inclusao": {
          "$gte": "2022-10-27T03:00:00.000Z",
          "$lte": "2022-10-27T03:00:00.000Z",
        },
        "data_inclusao_obito": {
          "$ne": null,
        },
      },
    },
    {
      "$addFields": {
        "data_inclusao": {
          "$toDate": "$data_inclusao",
        },
      },
    },
    {
      "$group": {
        "_id": {
          "month": {
            "$month": "$data_inclusao",
          },
          "year": {
            "$year": "$data_inclusao",
          },
        },
        "field": {
          "$avg": {
            "$toInt": "$idade",
          },
        },
      },
    },
    {
      "$project": {
        "_id": 0,
        "idade": {
          "$round": [
            "$field",
          ],
        },
        "month": "$_id.month",
        "year": "$_id.year",
      },
    },
    {
      "$sort": {
        "month": 1,
        "year": 1,
      },
    },
  ],
}
`;
